src/ucs/datastruct/array.inl

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_ARRAY_INL_
+#define UCS_ARRAY_INL_
+
+#include <ucs/datastruct/array.h>
+#include <ucs/sys/math.h>
+#include <ucs/debug/log.h>
+#include <ucs/debug/memtrack.h>
+
+
+/* Increase the array buffer length by this factor, whenever it needs to grow */
+#define UCS_ARRAY_GROW_FACTOR   2
+
+
+/**
+ * Define the functions of an array
+ *
+ * @param _name           Array name
+ * @param _index_type     Type of array's index
+ * @param _value_type     Type of array's values
+ * @param _storage_class  Storage class for array's functions (e.g 'static')
+ */
+#define UCS_ARRAY_FUNCS(_name, _index_type, _value_type, _storage_class) \
+    \
+    _storage_class void \
+    UCS_ARRAY_IDENTIFIER(_name, _init_dynamic)(ucs_array_t(_name) *array) \
+    { \
+        array->buffer   = NULL; \
+        array->length   = 0; \
+        array->capacity = 0; \
+    } \
+    \
+    _storage_class void \
+    UCS_ARRAY_IDENTIFIER(_name, _cleanup_dynamic)(ucs_array_t(_name) *array) \
+    { \
+        ucs_assert(!ucs_array_is_fixed(array)); \
+        ucs_free(array->buffer); \
+    } \
+    \
+    _storage_class ucs_status_t \
+    UCS_ARRAY_IDENTIFIER(_name, _grow)(ucs_array_t(_name) *array, \
+                                       _index_type min_capacity) \
+    { \
+        _index_type new_capacity; \
+        _value_type *new_buffer; \
+        size_t alloc_length; \
+        \
+        if (ucs_array_is_fixed(array)) { \
+            ucs_debug("cannot grow fixed array %s from %zu to %zu elems", \
+                      UCS_PP_MAKE_STRING(_name), (size_t)ucs_array_capacity(array), \
+                      (size_t)min_capacity); \
+            return UCS_ERR_NO_MEMORY; \
+        } \
+        \
+        new_capacity = ucs_max(array->capacity * UCS_ARRAY_GROW_FACTOR, \
+                               min_capacity); \
+        new_capacity = (new_capacity + ~UCS_ARRAY_CAP_MASK) & UCS_ARRAY_CAP_MASK; \
```

Comment:
```
new_capacity = ucs_array_capacity(new_capacity + ~UCS_ARRAY_CAP_MASK)
```

Reply:
IMO we can't to it, since ucs_array_capacity() accepts the array, but at this stage we still have not updated the array structure (since allocation can fail)